AnkiDroid/src/main/java/com/ichi2/anki/jsaddons/TgzPackageExtract.kt

```diff
@@ -0,0 +1,205 @@
+                var entry: TarArchiveEntry? = tarInputStream1.nextEntry as TarArchiveEntry
+
+                while (entry != null) {
+                    val outputFile = File(outputDir, entry.name)
+
+                    // Zip Slip Vulnerability https://snyk.io/research/zip-slip-vulnerability
+                    zipPathSafety(outputFile, outputDir)
+                    if (entry.isDirectory) {
+                        unTarDir(inputFile, outputDir, outputFile)
+                    } else {
+                        unTarFile(tarInputStream, entry, outputDir, outputFile)
+                    }
+
+                    entry = tarInputStream.nextEntry as? TarArchiveEntry?
+                }
+            }
+        }
+    }
+
+    /**
+     * UnTar file from archive using input stream, entry to output dir
+     * @param tarInputStream TarArchiveInputStream
+     * @param entry TarArchiveEntry
+     * @param outputDir Output directory
+     * @param outputFile Output file
+     * @throws IOException
+     */
+    @Throws(IOException::class)
+    private fun unTarFile(tarInputStream: TarArchiveInputStream, entry: TarArchiveEntry?, outputDir: File, outputFile: File) {
+        Timber.i("Creating output file %s.", outputFile.absolutePath)
+        val currentFile = File(outputDir, entry!!.name)
+        val parent = currentFile.parentFile // this line important otherwise FileNotFoundException
+
+        if (parent != null && !parent.exists()) {
+            if (!parent.mkdirs()) {
+                throw IOException(String.format("Couldn't create directory %s.", parent.absolutePath))
+            }
+        }
+
+        val outputFileStream: OutputStream = FileOutputStream(outputFile)
+        IOUtils.copy(tarInputStream, outputFileStream)
+        outputFileStream.close()
+    }
+
+    /**
+     * UnTar directory to output dir
+     * @param inputFile archive input file
+     * @param outputDir Output directory
+     * @param outputFile Output file
+     */
+    private fun unTarDir(inputFile: File, outputDir: File, outputFile: File) {
+        Timber.i("Untaring %s to dir %s.", inputFile.absolutePath, outputDir.absolutePath)
+        if (!outputFile.exists()) {
+            Timber.i("Attempting to create output directory %s.", outputFile.absolutePath)
+            check(outputFile.mkdirs()) { String.format("Couldn't create directory %s.", outputFile.absolutePath) }
+        }
+    }
+
+    // Zip Slip Vulnerability https://snyk.io/research/zip-slip-vulnerability
+    @Throws(ArchiveException::class, IOException::class)
+    private fun zipPathSafety(outputFile: File, destDirectory: File) {
+        val destDirCanonicalPath = destDirectory.canonicalPath
+        val outputFileCanonicalPath = outputFile.canonicalPath
+
+        if (!outputFileCanonicalPath.startsWith(destDirCanonicalPath)) {
+            throw ArchiveException(String.format("Entry is outside of the target dir: %s", outputFileCanonicalPath))
```

Comment:
I think this should be more strongly worded, but also should be pulled in to a string constant for translation since I think these errors will be exposed to users in future PRs

```suggestion
            throw ArchiveException(String.format("Malicious archive. Contains entry outside of the target dir: %s", outputFileCanonicalPath))
```